File: support/api-testsupport/src/main/java/com/hotels/styx/support/api/FakeHttpServer.java
/*
  Copyright (C) 2013-2024 Expedia Inc.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
package com.hotels.styx.support.api;

import com.github.tomakehurst.wiremock.WireMockServer;
import com.github.tomakehurst.wiremock.client.MappingBuilder;
import com.github.tomakehurst.wiremock.client.ResponseDefinitionBuilder;
import com.github.tomakehurst.wiremock.client.WireMock;
import com.github.tomakehurst.wiremock.common.HttpsSettings;
import com.github.tomakehurst.wiremock.core.WireMockConfiguration;
import com.github.tomakehurst.wiremock.matching.RequestPatternBuilder;
import com.github.tomakehurst.wiremock.matching.UrlPattern;

import static com.github.tomakehurst.wiremock.client.WireMock.configureFor;
import static com.github.tomakehurst.wiremock.client.WireMock.stubFor;
import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.wireMockConfig;

public final class FakeHttpServer {
    private final String appId;
    private final String originId;
    /*
             * The FakeHttpServer can be started as a HTTPS server or a bare HTTP server.
             *
             * The FakeHttpServer is built on top of WireMock, which also supports secure HTTPS protocol.
             * When WireMock is used as a HTTPS mock server, it still needs to be configured via
             * non-SSL HTTP protocol. For this reason, the FakeHttpServer uses two separate ports in SSL
             * mode: An 'adminPort' for configuring and assertions, and 'serverPort' for serving the mock
             * endpoints.
             *
             * When FakeHttpServer is started as a bare HTTP server, the adminPort and serverPort are
             * the same.
             *
             */
    private final HttpsSettings httpsSettings;
    private final int serverPort;
    private final int adminPort;
    private final WireMockServer server;

    static {
        System.setProperty("org.mortbay.log.class", "com.github.tomakehurst.wiremock.jetty.LoggerAdapter");
    }

    private FakeHttpServer(String appId, String originId, WireMockConfiguration wireMockConfiguration) {
        this.appId = appId;
        this.originId = originId;
        httpsSettings = wireMockConfiguration.httpsSettings();
        server = new WireMockServer(wireMockConfiguration);

        if (httpsSettings.enabled()) {
            this.adminPort = wireMockConfiguration.portNumber();
            this.serverPort = httpsSettings.port();
        } else {
            this.adminPort = wireMockConfiguration.portNumber();
            this.serverPort = wireMockConfiguration.portNumber();
        }
    }

    public FakeHttpServer(int port) {
        this("generic-app", "generic-app-" + port, wireMockConfig().port(port));
    }

    public static FakeHttpServer newHttpServer(int port) {
        return new FakeHttpServer(port);
    }

    public static FakeHttpServer newHttpServer(String appId, String originId, WireMockConfiguration wireMockConfiguration) {
        return new FakeHttpServer(appId, originId, wireMockConfiguration);
    }


    public FakeHttpServer start() {
        configureFor("localhost", adminPort);
        if (!server.isRunning()) {
            server.start();
        }
        return this;
    }

    public FakeHttpServer stub(UrlPattern urlPattern, ResponseDefinitionBuilder response) {
        configureFor("localhost", adminPort());
        stubFor(WireMock.get(urlPattern).willReturn(response));
        return this;
    }

    public FakeHttpServer stub(MappingBuilder mappingBuilder, ResponseDefinitionBuilder response) {
        configureFor("localhost", adminPort());
        stubFor(mappingBuilder.willReturn(response));
        return this;
    }

    public void verify(int count, RequestPatternBuilder builder) {
        configureFor("localhost", adminPort());
        WireMock.verify(count, builder);
    }

    public void verify(RequestPatternBuilder builder) {
        configureFor("localhost", adminPort());
        WireMock.verify(builder);
    }

    public FakeHttpServer stop() {
        configureFor("localhost", adminPort());
        if (server.isRunning()) {
            server.stop();
        }
        return this;
    }

    public FakeHttpServer reset() {
        configureFor("localhost", adminPort());
        WireMock.reset();
        return this;
    }

    public String appId() {
        return appId;
    }

    public String originId() {
        return originId;
    }

    public int port() {
        return serverPort;
    }

    public int adminPort() {
        return adminPort;
    }

    public boolean ssl() {

        return httpsSettings.enabled();
    }

    public FakeHttpServer setDelay(int delay) {
        configureFor("localhost", adminPort());
        WireMock.setGlobalFixedDelay(delay);
        return this;
    }

    public boolean isRunning() {
        return server.isRunning();
    }
}

File: system-tests/e2e-suite/src/test/scala/com/hotels/styx/support/backends/FakeHttpServer.scala
/*
  Copyright (C) 2013-2018 Expedia Inc.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
package com.hotels.styx.support.backends

import java.nio.charset.Charset

import com.github.tomakehurst.wiremock.client.WireMock._
import com.github.tomakehurst.wiremock.core.WireMockConfiguration
import com.hotels.styx.api.HttpHeaderNames.CONTENT_LENGTH
import com.hotels.styx.server.HttpsConnectorConfig
import com.hotels.styx.servers.MockOriginServer
import com.hotels.styx.support.server.UrlMatchingStrategies.urlStartingWith
import com.hotels.styx.support.server.{FakeHttpServer => JavaFakeHttpServer}
import com.hotels.styx.utils.StubOriginHeader.STUB_ORIGIN_INFO

import scala.collection.JavaConverters._

object FakeHttpServer {

  case class HttpsStartupConfig(httpsPort: Int = 0,
                                adminPort: Int = 0,
                                appId: String = "generic-app",
                                originId: String = "generic-app",
                                certificateFile: String = null,
                                certificateKeyFile: String = null,
                                protocols: Seq[String] = Seq("TLSv1.1", "TLSv1.2"),
                                cipherSuites: Seq[String] = Seq(),
                                sslProvider: String = "JDK"
                               ) {

    def start(): MockOriginServer = {

      var builder = new HttpsConnectorConfig.Builder()
      .sslProvider(sslProvider)
        .port(httpsPort)
        .protocols(protocols:_*)
        .cipherSuites(cipherSuites.toList.asJava)
      builder = if (certificateFile != null) builder.certificateFile(certificateFile) else builder
      builder = if (certificateKeyFile != null) builder.certificateFile(certificateKeyFile) else builder

      MockOriginServer.create(appId, originId, adminPort, builder.build()).start()
    }
  }

  case class HttpStartupConfig(port: Int = 0,
                               appId: String = "generic-app",
                               originId: String = "generic-app-01"
                              ) {
    private def asJava: WireMockConfiguration = {
      val wmConfig = if (port == 0) {
        new WireMockConfiguration().dynamicPort()
      } else {
        new WireMockConfiguration().port(port)
      }

      wmConfig.httpsPort(-1)
    }

    def start(): JavaFakeHttpServer = {
      val server = JavaFakeHttpServer.newHttpServer(appId, originId, this.asJava).start()
      println("server ports: " + server.adminPort() + " " + server.port())

      val response = s"Response From $appId:$originId, localhost:$port"

      server.stub(urlStartingWith("/"), aResponse
        .withStatus(200)
        .withHeader(CONTENT_LENGTH.toString, response.getBytes(Charset.defaultCharset()).size.toString)
        .withHeader(STUB_ORIGIN_INFO.toString, s"{appId.toUpperCase}-$originId")
        .withBody(response))

      server
    }
  }

  private object WireMockDefaults {
    val https = new WireMockConfiguration().httpsSettings()
  }

}
